From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Mon, 29 Nov 2021 22:20:14 +1000
Subject: [PATCH] Disable Paper timings by default


diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index 456595e4b7e0c7f50617aa2694b0d2dfc368ab81..80804b54f6cc7c7b26db6cc2c6c33b16eaec1c3f 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -52,6 +52,8 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     public class Timings extends ConfigurationPart.Post {
         public boolean enabled = true;
+
+        public boolean reallyEnableTimings = true;
         public boolean verbose = true;
         public String url = "https://timings.aikar.co/";
         public boolean serverNamePrivacy = false;
@@ -65,6 +67,12 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         @Override
         public void postProcess() {
+            if (enabled && !reallyEnableTimings) {
+                org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING, "[Airplane] To improve performance, timings have been disabled by default");
+                org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING, "[Airplane] You can still use timings by using /timings on, but they will not start on server startup unless you set timings.really-enabled to true in paper.yml");
+                org.bukkit.Bukkit.getLogger().log(java.util.logging.Level.WARNING, "[Airplane] If you would like to disable this message, either set timings.really-enabled to true or timings.enabled to false.");
+            }
+            enabled = reallyEnableTimings;
             MinecraftTimings.processConfig(this);
         }
     }
