From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 28 Nov 2021 23:37:01 +1000
Subject: [PATCH] Tick attack strength for external players


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 36c1b695c188e10a765419b63bfb47839e8c1a9f..69077329af2e2adbe061d38283ddde19c92687cf 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -307,6 +307,12 @@ public abstract class Player extends LivingEntity {
             this.setPos(d0, this.getY(), d1);
         }
 
+    // MultiPaper start
+        tickAttackStrength();
+        tickAttackStrengthAfter();
+    }
+
+    public void tickAttackStrength() {
         ++this.attackStrengthTicker;
         ItemStack itemstack = this.getMainHandItem();
 
@@ -317,6 +323,10 @@ public abstract class Player extends LivingEntity {
 
             this.lastItemInMainHand = itemstack.copy();
         }
+    }
+
+    public void tickAttackStrengthAfter() {
+    // MultiPaper end
 
         this.turtleHelmetTick();
         this.cooldowns.tick();
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index dd54000a43060736a1515e0a18abc51ebc356e63..d5c2407ce3e3ad1fa55d616723fa40a300ffce29 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -91,6 +91,8 @@ public class MultiPaper {
                 // Copied from Paper end
 
                 externalPlayer.checkInsideBlocks();
+
+                externalPlayer.tickAttackStrength();
             }
         }
 
