From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Sun, 27 Feb 2022 11:24:14 +1000
Subject: [PATCH] Update external player spam counters


diff --git a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
index 6c4143be21cdb9cc6a4cb20de5bdec06c3068c76..6bfc10236ad8ce288ac2a50eb41df7355f83486a 100644
--- a/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/src/main/java/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -399,20 +399,23 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         // Paper end
 
         this.server.getProfiler().pop();
-        // CraftBukkit start
-        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
-        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
-        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
-        /* Use thread-safe field access instead
-        if (this.chatSpamTickCount > 0) {
-            --this.chatSpamTickCount;
-        }
-        */
-        // CraftBukkit end
-
-        if (this.dropSpamTickCount > 0) {
-            --this.dropSpamTickCount;
-        }
+// MultiPaper start - move to reduceSpamCounters()
+//        // CraftBukkit start
+//        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
+//        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
+//        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
+//        /* Use thread-safe field access instead
+//        if (this.chatSpamTickCount > 0) {
+//            --this.chatSpamTickCount;
+//        }
+//        */
+//        // CraftBukkit end
+//
+//        if (this.dropSpamTickCount > 0) {
+//            --this.dropSpamTickCount;
+//        }
+        reduceSpamCounters();
+// MultiPaper end
 
         if (this.player.getLastActionTime() > 0L && this.server.getPlayerIdleTimeout() > 0 && Util.getMillis() - this.player.getLastActionTime() > (long) (this.server.getPlayerIdleTimeout() * 1000 * 60) && !this.player.wonGame) { // Paper - Prevent AFK kick while watching end credits.
             this.player.resetLastActionTime(); // CraftBukkit - SPIGOT-854
@@ -431,6 +434,25 @@ public class ServerGamePacketListenerImpl implements ServerPlayerConnection, Ser
         this.lastGoodZ = this.player.getZ();
     }
 
+    // MultiPaper start
+    public void reduceSpamCounters() {
+        // CraftBukkit start
+        for (int spam; (spam = this.chatSpamTickCount.get()) > 0 && !this.chatSpamTickCount.compareAndSet(spam, spam - 1); ) ;
+        if (tabSpamLimiter.get() > 0) tabSpamLimiter.getAndDecrement(); // Paper - split to seperate variable
+        if (recipeSpamPackets.get() > 0) recipeSpamPackets.getAndDecrement(); // Paper
+        /* Use thread-safe field access instead
+        if (this.chatSpamTickCount > 0) {
+            --this.chatSpamTickCount;
+        }
+        */
+        // CraftBukkit end
+
+        if (this.dropSpamTickCount > 0) {
+            --this.dropSpamTickCount;
+        }
+    }
+    // MultiPaper end
+
     @Override
     public Connection getConnection() {
         return this.connection;
diff --git a/src/main/java/puregero/multipaper/MultiPaper.java b/src/main/java/puregero/multipaper/MultiPaper.java
index dbc0ed2a8d91c266cb13decd5cef20307cd04f5c..62f9527c343a467f73e8cd0425647cb0d9d3ba4f 100644
--- a/src/main/java/puregero/multipaper/MultiPaper.java
+++ b/src/main/java/puregero/multipaper/MultiPaper.java
@@ -101,6 +101,8 @@ public class MultiPaper {
             }
 
             player.syncExperience();
+
+            player.connection.reduceSpamCounters();
         }
 
         for (ExternalServer server : getConnection().getServersMap().values()) {
