From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Tue, 8 Mar 2022 00:37:35 +1000
Subject: [PATCH] Only drop leads on the local server


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 49671da87f91a1103818df514c4f8dd76c6f6f3d..3a490a964c7c5939601d6c0ab2a68446a5a20058 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -84,6 +84,7 @@ import org.bukkit.event.entity.EntityTargetEvent;
 import org.bukkit.event.entity.EntityTransformEvent;
 import org.bukkit.event.entity.EntityUnleashEvent;
 import org.bukkit.event.entity.EntityUnleashEvent.UnleashReason;
+import puregero.multipaper.MultiPaper;
 // CraftBukkit end
 
 public abstract class Mob extends LivingEntity {
@@ -1436,7 +1437,7 @@ public abstract class Mob extends LivingEntity {
         if (this.leashHolder != null) {
             this.leashHolder = null;
             this.leashInfoTag = null;
-            if (!this.level.isClientSide && dropItem) {
+            if (!this.level.isClientSide && dropItem && MultiPaper.isChunkLocal(level.getChunkIfLoaded(blockPosition))) { // MultiPaper - only drop leads on the local server
                 this.forceDrops = true; // CraftBukkit
                 this.spawnAtLocation((ItemLike) Items.LEAD);
                 this.forceDrops = false; // CraftBukkit
