From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 13 Apr 2022 16:54:41 +1000
Subject: [PATCH] Don't compare items if they're the same item


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index 36c1b695c188e10a765419b63bfb47839e8c1a9f..4e1372337279584b862e4491a742701f4a65a294 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -310,12 +310,12 @@ public abstract class Player extends LivingEntity {
         ++this.attackStrengthTicker;
         ItemStack itemstack = this.getMainHandItem();
 
-        if (!ItemStack.matches(this.lastItemInMainHand, itemstack)) {
+        if (this.lastItemInMainHand != itemstack && !ItemStack.matches(this.lastItemInMainHand, itemstack)) { // MultiPaper - don't compare items if they're the same item
             if (!ItemStack.isSameIgnoreDurability(this.lastItemInMainHand, itemstack)) {
                 this.resetAttackStrengthTicker();
             }
 
-            this.lastItemInMainHand = itemstack.copy();
+            this.lastItemInMainHand = itemstack; // MultiPaper - don't compare items if they're the same item
         }
 
         this.turtleHelmetTick();
